enhance docs for eks (#721)
localstack/docs: content/en/user-guide/aws/elastic-kubernetes-service/index.md

---
title: "Elastic Kubernetes Service (EKS)"
linkTitle: "Elastic Kubernetes Service (EKS)"
description: >
  Get started with Elastic Kubernetes Service (EKS) on LocalStack
---

## Introduction

> Elastic Kubernetes Service (EKS) is a managed Kubernetes service that makes it easy to run Kubernetes on AWS without installing, operating, and maintaining your own Kubernetes control plane or worker nodes. Kubernetes is an open-source system for automating containerised applications' deployment, scaling, and management.

Reviewer: nit: we usually use American (not British) spelling for our docs:
```suggestion
Elastic Kubernetes Service (EKS) is a managed Kubernetes service that makes it easy to run Kubernetes on AWS without installing, operating, and maintaining your own Kubernetes control plane or worker nodes. Kubernetes is an open-source system for automating containerized applications' deployment, scaling, and management.
```

LocalStack supports Elastic Kubernetes Service via the Pro/Team offering, allowing you to use the EKS APIs in your local environment to spin up embedded Kubernetes clusters in your local Docker engine or use an existing Kubernetes installation you can access from your local machine (defined in `$HOME/.kube/config`). The supported APIs are available on our [API coverage page](https://docs.localstack.cloud/references/coverage/coverage_eks/), which provides information on the extent of EKS's integration with LocalStack.

## Getting started

This guide is designed for users new to Elastic Kubernetes Service and assumes basic knowledge of the AWS CLI and our [`awslocal`](https://github.com/localstack/awscli-local) wrapper script. To interact with the Kubernetes cluster, you should also install [`kubectl`](https://kubernetes.io/docs/tasks/tools/).

Start your LocalStack container using your preferred method. We will demonstrate how you can auto-install an embedded Kubernetes cluster, configure ingress, and deploy a sample service with ECR.

### Create an embedded Kubernetes cluster

> The default approach to create Kubernetes clusters using the local EKS API is by setting up an embedded [k3d](https://k3d.io/) kube cluster within Docker. LocalStack seamlessly manages the download and installation process, making it hassle-free for users. In most cases, the installation is automatic, eliminating the need for any manual customizations.

Reviewer: tiny nit:
```suggestion
The default approach for creating Kubernetes clusters using the local EKS API is by setting up an embedded [k3d](https://k3d.io/) kube cluster within Docker. LocalStack seamlessly manages the download and installation process, making it hassle-free for users. In most cases, the installation is automatic, eliminating the need for any manual customizations.
```

A new cluster can be created using the following command:

You can create a new cluster using the [`CreateCluster`](https://docs.aws.amazon.com/eks/latest/APIReference/API_CreateCluster.html) API. Run the following command:

{{< command >}}
$ awslocal eks create-cluster \

>   --name fancier-cluster \

Reviewer: nit: "fancier-*" seems a bit unexpected here. It makes sense, when looking at the `fancier-nginx` pod further below, but wondering if we should err on the "unfancy" side here :) and use typical test names like "cluster1", "pod1, etc.. (not critical at all, though - feel free to merge as-is as well 👍 )

f05770ec8523   rancher/k3s:v1.21.5-k3s2       "/bin/k3s server --t…"   1 minute ago   Up 1 minute
...
{{</ command >}}

After successfully creating and initializing the cluster, we can easily find the server endpoint, using the [`DescribeCluster`](https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeCluster.html) API. Run the following command:

{{< command >}}
$ awslocal eks describe-cluster --name fancier-cluster
{
    "cluster": {
        "name": "fancier-cluster",
        "arn": "arn:aws:eks:us-east-1:000000000000:cluster/fancier-cluster",
        "createdAt": "2022-04-13T17:12:39.738000+02:00",
        "endpoint": "https://localhost.localstack.cloud:4511",
        "roleArn": "arn:aws:iam::000000000000:role/eks-role",
        "resourcesVpcConfig": {},
        "identity": {
            "oidc": {
                "issuer": "https://localhost.localstack.cloud/eks-oidc"
            }
        },
        "status": "ACTIVE",
        "certificateAuthority": {
            "data": "..."
        },
        "clientRequestToken": "d188f578-b353-416b-b309-5d8c76ecc4e2"
    }
}
{{</ command >}}

### Utilizing ECR Images within EKS

> You can now use ECR (Amazon Elastic Container Registry) images within your EKS (Amazon Elastic Kubernetes Service) environment.

Reviewer: nit: no need to re-introduce EKS, which has already been defined above..

```suggestion
You can now use ECR (Elastic Container Registry) images within your EKS environment.
```

{{< alert title="Notes" >}}
In this section, we assume that `localhost.localstack.cloud` resolves in your environment, and LocalStack is connected to a non-default bridge network. For more information, refer to the article about [DNS rebind protection]({{< ref "dns-server#dns-rebind-protection" >}}).

If the domain `localhost.localstack.cloud` does not resolve on your host, you can still proceed without setting `HOSTNAME_EXTERNAL`. However it is suggested to use `localhost.localstack.cloud` as the registry in your pod configuration. 

LocalStack will take care of the DNS resolution of `localhost.localstack.cloud` within ECR itself, allowing you to use the `localhost:<port>/<repository_name>` URI for tagging and pushing the image on your host.
{{< / alert >}}

Once you have configured this correctly, you can seamlessly use your ECR image within EKS as expected.

#### Deploying a sample application from an ECR image

> To showcase this behavior, let's go through a concise step-by-step guide that will lead us to the successful pulling of an image from ECR. For the purpose of this guide, we will retag the `nginx` image to be pushed to ECR under a different name and then utilize it for a pod configuration.

Reviewer: ```suggestion
To showcase this behavior, let's go through a concise step-by-step guide that will lead us to the successful pulling of an image from local ECR. For the purpose of this guide, we will retag the `nginx` image to be pushed to a local ECR repository under a different name, and then utilize it for a pod configuration.
```

{{< / command >}}

{{< alert title="Note">}}
When creating an ECR repository, a port from the [external service port range]({{< ref "external-ports" >}}) is dynamically assigned. As a result, the port can differ from the static value `4510` used in the examples below.

To ensure the correct URL and port, it's important to use the `repositoryUrl` obtained from the `create-repository` request. This ensures that you have the accurate endpoint to access the repository.
{{< /alert >}}

You can now pull the `nginx` image from Docker Hub using the `docker` CLI:

{{< command >}}
$ docker pull nginx
{{< / command >}}

You can further tag the image to be pushed to ECR:

{{< command >}}
$ docker tag nginx localhost.localstack.cloud:4510/fancier-nginx
{{< / command >}}

> Finally, you can push the image to ECR:

Reviewer: ```suggestion
Finally, you can push the image to local ECR:
```

{{< command >}}
$ docker push localhost.localstack.cloud:4510/fancier-nginx
{{< / command >}}

Now, let us set up the EKS cluster using the image pushed to local ECR.

Next, we can configure `kubectl` to use the EKS cluster, using the [`UpdateKubeconfig`](https://docs.aws.amazon.com/eks/latest/APIReference/API_UpdateClusterConfig.html) API. Run the following command:

{{< command >}}

> $ awslocal eks update-kubeconfig \

Reviewer: nit: line break seems a bit weird - maybe break up the command into two separate lines (each starting with `$`)..?

EOF
{{< / command >}}

You can now describe the pod to see if the image was pulled successfully:

{{< command >}}

> kubectl describe pod fancier-nginx

Reviewer: Would be good to add the `$` indicator to all commands, for consistency.. 👍 
```suggestion
$ kubectl describe pod fancier-nginx
```

{{< / command >}}

In the events, we can see that the pull from ECR was successful:

```bash
  Normal  Pulled     10s   kubelet            Successfully pulled image "localhost.localstack.cloud:4510/fancier-nginx:latest" in 2.412775896s
```

### Configuring an Ingress for your services

To make an EKS service externally accessible, it is necessary to create an Ingress configuration, which exposes the service on a specific path to the load balancer.

> For our sample deployment, we can create an nginx Kubernetes service by applying the following configuration:

Reviewer: ```suggestion
For our sample deployment, we can create an `nginx` Kubernetes service by applying the following configuration:
```

{{< / command >}}

{{< alert title="Note" >}}
You can customize the Load Balancer port by configuring `EKS_LOADBALANCER_PORT` in your environment.
{{< /alert >}}

### Enabling HTTPS with local SSL/TLS certificate for the Ingress

> To enable HTTPS for your endpoints, you can configure Kubernetes to use SSL/TLS with your [certificate for local domain names](https://github.com/localstack/localstack-artifacts/blob/master/local-certs/server.key) `*.localhost.localstack.cloud`.

Reviewer: "your certificate" doesn't seem fully accurate in this context. I'd either leave it as "our certificate" (as we're providing it), or simply use "the certificate".

}
{{< / command >}}

To interact with your Kubernetes cluster, configure your Kubernetes client (such as `kubectl` or other SDKs) to point to the `endpoint` provided in the `create-cluster` output mentioned earlier. However, depending on whether you're calling the Kubernetes API from your local machine or from within a Lambda function, you might need to use different endpoint URLs.

> For local machine interactions, use `https://localhost:6443` as the endpoint URL. If you are accessing the Kubernetes API from within a Lambda function, you should use `https://172.17.0.1:6443` as the endpoint URL.

Reviewer: Maybe better use `host.docker.internal` instead of the hardcoded IP address, or at least mention something like "... (assuming that `172.17.0.1` is the IP address of the Docker network bridge)".

In the provided example, we define routing rules for two local endpoints. The first rule directs traffic to a service named `service-1`, accessible under the path `/v1`. Similarly, the second rule points to a service named `service-2`, also accessible under the same path `/v1`.

This approach enables us to access the two distinct services using the same path and port number, but with different host names. This host-based routing mechanism ensures that each service is uniquely identified based on its designated host name, allowing for a uniform and organized way of accessing multiple services within the EKS cluster.

{{< command >}}
$ curl http://eks-service-1.localhost.localstack.cloud:8081/v1
... [output of service 1]
$ curl http://eks-service-2.localhost.localstack.cloud:8081/v1
... [output of service 2]
{{< /command >}}

It is important to note that the host names `eks-service-1.localhost.localstack.cloud` and `eks-service-2.localhost.localstack.cloud` both resolve to `127.0.0.1` (localhost). Consequently, you can utilize them to communicate with your service endpoints and distinguish between different services within the Kubernetes load balancer.

> However, it might encounter issues in scenarios where you intend to run your Load Balancer (LB) on standard ports such as 80/443 since some of these ports could already be occupied on your local machine. For instance, by default, LocalStack allocates port 443 to expose APIs via the HTTPS endpoint (`https://localhost.localstack.cloud`). Hence, it's crucial to ensure that you expose your LB on a custom, non-standard port to prevent conflicts.

Reviewer: ```suggestion
However, it might encounter issues in scenarios where you intend to run your Load Balancer (LB) on standard ports such as 80/443 since some of these ports may already be occupied on your local machine. For instance, by default, LocalStack allocates port 443 to expose APIs via the HTTPS endpoint (`https://localhost.localstack.cloud`). Hence, it's crucial to ensure that you expose your LB on a custom, non-standard port to prevent conflicts.
```
